sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java

```diff
@@ -33,32 +36,65 @@
 public class PCollectionConsumerRegistry {
 
   private ListMultimap<String, FnDataReceiver<WindowedValue<?>>> pCollectionIdsToConsumers;
+  private Map<String, ElementCountFnDataReceiver> pCollectionIdsToWrappedConsumer;
 
   public PCollectionConsumerRegistry() {
     pCollectionIdsToConsumers = ArrayListMultimap.create();
+    pCollectionIdsToWrappedConsumer = new HashMap<String, ElementCountFnDataReceiver>();
   }
 
-  public <T> FnDataReceiver<WindowedValue<T>> registerAndWrap(
-      String pCollectionId, FnDataReceiver<WindowedValue<T>> consumer) {
-    // TODO(ajamato): Test multiple consumers of the same pcollection before merging to master.
-    FnDataReceiver<WindowedValue<T>> wrappedReceiver =
-        new ElementCountFnDataReceiver<T>(consumer, pCollectionId);
-    pCollectionIdsToConsumers.put(pCollectionId, (FnDataReceiver) wrappedReceiver);
-    return wrappedReceiver;
+  public <T> void register(String pCollectionId, FnDataReceiver<WindowedValue<T>> consumer) {
+    // Just save these consumers for now, but package them up later with an
+    // ElementCountFnDataReceiver and possibly a MultiplexingFnDataReceiver
+    // if there are multiple consumers.
+    ElementCountFnDataReceiver wrappedConsumer =
+        pCollectionIdsToWrappedConsumer.getOrDefault(pCollectionId, null);
+    if (wrappedConsumer != null) {
+      throw new RuntimeException(
```

Comment:
We might split this to builder and implementation. This will remove the option for user to invoke calls in wrong order.